Panic in Signature Verification for Namespace Creation with Empty Endorsements

## Description
The signature verifier panics when processing a namespace creation transaction that uses a key-based policy but has no endorsements.

Version: v0.1.8

## Steps to Reproduce
1. Configure a namespace with a key-based verification policy (currently still using metaNamespaceKey as of v0.1.8)
2. Send a namespace creation transaction without endorsements
3. The verifier service crashes with a panic

## Expected Behavior
The verifier should return a descriptive error message indicating that endorsements are required for key-based verification.

## Actual Behavior
The service panics with:
```
panic: runtime error: index out of range [0] with length 0

goroutine 86 [running]:
github.com/hyperledger/fabric-x-committer/utils/signature.(*keyVerifier).Verify(...)
	/utils/signature/verify.go:119 +0xb8
```

## Root Cause
In `utils/signature/verify.go:119`, the `keyVerifier.Verify` method accesses `endorsements[0]` without checking if the slice is empty:

```go
func (v *keyVerifier) Verify(data []byte, endorsements []*applicationpb.EndorsementWithIdentity) error {
	digest := sha256.Sum256(data)
	return v.verifyDigest(digest[:], endorsements[0].Endorsement)  // Line 119 - No bounds check
}
```

## Suggested Fix
Add validation before accessing the endorsements slice:

```go
func (v *keyVerifier) Verify(data []byte, endorsements []*applicationpb.EndorsementWithIdentity) error {
	if len(endorsements) == 0 {
		return errors.New("some meaningful error")
	}
	digest := sha256.Sum256(data)
	return v.verifyDigest(digest[:], endorsements[0].Endorsement)
}
```

## Impact
- **Severity**: High - causes service crash
- **Component**: Signature Verification Service
- **Affected File**: `utils/signature/verify.go`

--- a/integration/test/namespace_lifecycle_test.go
+++ b/integration/test/namespace_lifecycle_test.go
@@ -16,6 +16,7 @@ import (
 	"github.com/stretchr/testify/require"
 	"google.golang.org/protobuf/proto"
 
+	"github.com/hyperledger/fabric-x-committer/api/servicepb"
 	"github.com/hyperledger/fabric-x-committer/integration/runner"
 )
 
@@ -137,4 +138,25 @@ func TestCreateUpdateNamespace(t *testing.T) {
 			c.MakeAndSendTransactionsToOrderer(t, tt.txs, tt.expected)
 		})
 	}
+
+	t.Run("empty endorsements", func(t *testing.T) {
+		tx := &applicationpb.Tx{
+			Namespaces: []*applicationpb.TxNamespace{{
+				NsId:      "1",
+				NsVersion: 2,
+				BlindWrites: []*applicationpb.Write{{
+					Key:   []byte("key5"),
+					Value: []byte("value5"),
+				}},
+			}},
+			Endorsements: []*applicationpb.Endorsements{{
+				EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{},
+			}},
+		}
+		lgTx := c.TxBuilder.MakeTx(tx)
+		c.SendTransactionsToOrderer(t,
+			[]*servicepb.LoadGenTx{lgTx},
+			[]committerpb.Status{committerpb.Status_MALFORMED_MISSING_SIGNATURE},
+		)
+	})
 }

--- a/service/sidecar/mapping.go
+++ b/service/sidecar/mapping.go
@@ -243,8 +243,8 @@ func verifyTxForm(tx *applicationpb.Tx) committerpb.Status {
 	if len(tx.Namespaces) == 0 {
 		return committerpb.Status_MALFORMED_EMPTY_NAMESPACES
 	}
-	if len(tx.Namespaces) != len(tx.Endorsements) {
-		return committerpb.Status_MALFORMED_MISSING_SIGNATURE
+	if status := checkEndorsements(tx); status != statusNotYetValidated {
+		return status
 	}
 
 	nsIDs := make(map[string]any, len(tx.Namespaces))
@@ -265,6 +265,27 @@ func verifyTxForm(tx *applicationpb.Tx) committerpb.Status {
 			}
 		}
 		nsIDs[ns.NsId] = nil
+	}
+	return statusNotYetValidated
+}
+
+func checkEndorsements(tx *applicationpb.Tx) committerpb.Status {
+	if len(tx.Namespaces) != len(tx.Endorsements) {
+		return committerpb.Status_MALFORMED_MISSING_SIGNATURE
+	}
+	for _, e := range tx.Endorsements {
+		if e == nil || len(e.EndorsementsWithIdentity) == 0 {
+			return committerpb.Status_MALFORMED_MISSING_SIGNATURE
+		}
+		for _, ei := range e.EndorsementsWithIdentity {
+			if ei == nil || len(ei.Endorsement) == 0 {
+				return committerpb.Status_MALFORMED_MISSING_SIGNATURE
+			}
+		}
+		// Note: we do not validate the Identity field here because the sidecar does not know
+		// whether the namespace uses an MSP rule or a threshold rule for endorsement.
+		// Threshold rules do not require an identity. Identity validation is left to the
+		// signature verifier, which has the policy context to determine what is required.
 	}
 	return statusNotYetValidated
 }

--- a/service/sidecar/sidecar_test.go
+++ b/service/sidecar/sidecar_test.go
@@ -750,6 +750,7 @@ func makeValidTx(t *testing.T, chanID string) *servicepb.LoadGenTx {
 			NsVersion:   0,
 			BlindWrites: []*applicationpb.Write{{Key: utils.MustRead(rand.Reader, 32)}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	})
 }
 

--- a/service/sidecar/test_exports.go
+++ b/service/sidecar/test_exports.go
@@ -66,19 +66,30 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 		expectedStatuses = append(expectedStatuses, expected)
 	}
 
+	// placeholderEndorsements returns dummy endorsements when no TxEndorser is available
+	// (e.g., sidecar unit tests that only check form validation). When a TxEndorser is
+	// present (e.g., integration tests), it returns nil so MakeTx uses the TxEndorser
+	// to produce real signatures that pass verification.
+	placeholderEndorsements := func(n int) []*applicationpb.Endorsements {
+		if txb.TxEndorser == nil {
+			return dummyEndorsements(n)
+		}
+		return nil
+	}
+
 	validTxNamespaces := []*applicationpb.TxNamespace{{
 		NsId:        "1",
 		NsVersion:   0,
 		BlindWrites: []*applicationpb.Write{{Key: []byte("k1")}},
 	}}
-	validTx := txb.MakeTx(&applicationpb.Tx{Namespaces: validTxNamespaces})
+	validTx := txb.MakeTx(&applicationpb.Tx{Namespaces: validTxNamespaces, Endorsements: placeholderEndorsements(1)})
 
 	add(committerpb.Status_COMMITTED, validTx)
 	add(committerpb.Status_REJECTED_DUPLICATE_TX_ID, txb.MakeTxWithID(validTx.Id, &applicationpb.Tx{
-		Namespaces: validTxNamespaces,
+		Namespaces: validTxNamespaces, Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_MISSING_TX_ID, txb.MakeTxWithID("", &applicationpb.Tx{
-		Namespaces: validTxNamespaces,
+		Namespaces: validTxNamespaces, Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_EMPTY_NAMESPACES, txb.MakeTx(&applicationpb.Tx{}))
 	add(committerpb.Status_MALFORMED_EMPTY_NAMESPACES, txb.MakeTx(&applicationpb.Tx{
@@ -92,18 +103,40 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 		Namespaces:   validTxNamespaces,
 		Endorsements: make([]*applicationpb.Endorsements, 2), // Too many signatures.
 	}))
+	add(committerpb.Status_MALFORMED_MISSING_SIGNATURE, txb.MakeTx(&applicationpb.Tx{
+		Namespaces: validTxNamespaces,
+		Endorsements: []*applicationpb.Endorsements{{
+			EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{}, // Empty inner slice.
+		}},
+	}))
+	add(committerpb.Status_MALFORMED_MISSING_SIGNATURE, txb.MakeTx(&applicationpb.Tx{
+		Namespaces: validTxNamespaces,
+		Endorsements: []*applicationpb.Endorsements{{
+			EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{nil}, // Nil endorsement entry.
+		}},
+	}))
+	add(committerpb.Status_MALFORMED_MISSING_SIGNATURE, txb.MakeTx(&applicationpb.Tx{
+		Namespaces: validTxNamespaces,
+		Endorsements: []*applicationpb.Endorsements{{
+			EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{{
+				Endorsement: nil, // Nil signature bytes.
+			}},
+		}},
+	}))
 	add(committerpb.Status_MALFORMED_NO_WRITES, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
 			NsId:      "1",
 			NsVersion: 0,
 			ReadsOnly: []*applicationpb.Read{{Key: []byte("k1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_NAMESPACE_ID_INVALID, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{
 			// namespace id is invalid.
 			{NsId: "//", BlindWrites: validTxNamespaces[0].BlindWrites},
 		},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_NAMESPACE_ID_INVALID, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{
@@ -115,6 +148,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 				ReadWrites: []*applicationpb.ReadWrite{{Key: []byte("/\\")}},
 			},
 		},
+		Endorsements: dummyEndorsements(2),
 	}))
 	add(committerpb.Status_MALFORMED_NAMESPACE_POLICY_INVALID, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{
@@ -128,9 +162,11 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 				}},
 			},
 		},
+		Endorsements: dummyEndorsements(2),
 	}))
 	add(committerpb.Status_MALFORMED_DUPLICATE_NAMESPACE, txb.MakeTx(&applicationpb.Tx{
-		Namespaces: []*applicationpb.TxNamespace{validTxNamespaces[0], validTxNamespaces[0]},
+		Namespaces:   []*applicationpb.TxNamespace{validTxNamespaces[0], validTxNamespaces[0]},
+		Endorsements: dummyEndorsements(2),
 	}))
 	add(committerpb.Status_COMMITTED, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -142,6 +178,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 				Value: defaultNsValidPolicy(),
 			}},
 		}},
+		Endorsements: placeholderEndorsements(1),
 	}))
 	add(committerpb.Status_COMMITTED, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{
@@ -157,6 +194,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 				}},
 			},
 		},
+		Endorsements: placeholderEndorsements(2),
 	}))
 	add(committerpb.Status_MALFORMED_NAMESPACE_POLICY_INVALID, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -168,6 +206,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 				Value:   defaultNsInvalidPolicy(), // invalid policy.
 			}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_BLIND_WRITES_NOT_ALLOWED, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{
@@ -182,6 +221,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 				}},
 			},
 		},
+		Endorsements: dummyEndorsements(2),
 	}))
 	add(committerpb.Status_MALFORMED_EMPTY_KEY, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -190,6 +230,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			ReadsOnly:  []*applicationpb.Read{{Key: nil}},
 			ReadWrites: []*applicationpb.ReadWrite{{Key: []byte("1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_EMPTY_KEY, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -197,6 +238,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			NsVersion:  0,
 			ReadWrites: []*applicationpb.ReadWrite{{Key: nil}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_EMPTY_KEY, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -204,6 +246,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			NsVersion:   0,
 			BlindWrites: []*applicationpb.Write{{Key: nil}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_DUPLICATE_KEY_IN_READ_WRITE_SET, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -212,6 +255,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			ReadsOnly:  []*applicationpb.Read{{Key: []byte("key1")}, {Key: []byte("key1")}},
 			ReadWrites: []*applicationpb.ReadWrite{{Key: []byte("1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_DUPLICATE_KEY_IN_READ_WRITE_SET, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -219,6 +263,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			NsVersion:  0,
 			ReadWrites: []*applicationpb.ReadWrite{{Key: []byte("key1")}, {Key: []byte("key1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_DUPLICATE_KEY_IN_READ_WRITE_SET, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -226,6 +271,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			NsVersion:   0,
 			BlindWrites: []*applicationpb.Write{{Key: []byte("key1")}, {Key: []byte("key1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_DUPLICATE_KEY_IN_READ_WRITE_SET, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -234,6 +280,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			ReadsOnly:  []*applicationpb.Read{{Key: []byte("key1")}},
 			ReadWrites: []*applicationpb.ReadWrite{{Key: []byte("key1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	add(committerpb.Status_MALFORMED_DUPLICATE_KEY_IN_READ_WRITE_SET, txb.MakeTx(&applicationpb.Tx{
 		Namespaces: []*applicationpb.TxNamespace{{
@@ -242,6 +289,7 @@ func MalformedTxTestCases(txb *workload.TxBuilder) (
 			ReadWrites:  []*applicationpb.ReadWrite{{Key: []byte("key1")}},
 			BlindWrites: []*applicationpb.Write{{Key: []byte("key1")}},
 		}},
+		Endorsements: dummyEndorsements(1),
 	}))
 	return txs, expectedStatuses
 }
@@ -253,4 +301,16 @@ func defaultNsInvalidPolicy() []byte {
 func defaultNsValidPolicy() []byte {
 	_, verificationKey := sigtest.NewKeyPair(signature.Ecdsa)
 	return protoutil.MarshalOrPanic(policy.MakeECDSAThresholdRuleNsPolicy(verificationKey))
+}
+
+func dummyEndorsements(n int) []*applicationpb.Endorsements {
+	e := make([]*applicationpb.Endorsements, n)
+	for i := range e {
+		e[i] = &applicationpb.Endorsements{
+			EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{{
+				Endorsement: []byte("placeholder"),
+			}},
+		}
+	}
+	return e
 }

--- a/utils/signature/verify.go
+++ b/utils/signature/verify.go
@@ -132,6 +132,9 @@ func (*keyVerifier) UpdateIdentities(msp.IdentityDeserializer) error {
 
 // Verify evaluates the endorsements against the data for keyVerifier.
 func (v *keyVerifier) Verify(data []byte, endorsements []*applicationpb.EndorsementWithIdentity) error {
+	if len(endorsements) == 0 {
+		return errors.New("no endorsements provided for key-based verification")
+	}
 	digest := sha256.Sum256(data)
 	return v.verifyDigest(digest[:], endorsements[0].Endorsement)
 }
@@ -166,8 +169,14 @@ func (v *channelPolicyVerifier) Verify(data []byte, endorsements []*applicationp
 
 // evaluateEndorsements constructs SignedData from endorsements and evaluates them against a policy.
 func evaluateEndorsements(p policies.Policy, data []byte, endorsements []*applicationpb.EndorsementWithIdentity) error {
+	if len(endorsements) == 0 {
+		return errors.New("no endorsements provided for MSP-based verification")
+	}
 	signedData := make([]*protoutil.SignedData, len(endorsements))
 	for i, s := range endorsements {
+		if s.Identity == nil {
+			return errors.New("endorsement has nil identity")
+		}
 		signedData[i] = &protoutil.SignedData{
 			Data:      data,
 			Identity:  s.Identity,

--- a/utils/signature/verify_test.go
+++ b/utils/signature/verify_test.go
@@ -11,6 +11,7 @@ import (
 	"testing"
 
 	"github.com/hyperledger/fabric-x-common/api/applicationpb"
+	"github.com/hyperledger/fabric-x-common/api/msppb"
 	"github.com/hyperledger/fabric-x-common/common/cauthdsl"
 	"github.com/hyperledger/fabric-x-common/common/policydsl"
 	"github.com/hyperledger/fabric-x-common/msp"
@@ -44,6 +45,53 @@ func TestNsVerifierThresholdRule(t *testing.T) {
 	require.NoError(t, err)
 	tx1.Endorsements = []*applicationpb.Endorsements{endorsement}
 	require.NoError(t, nsVerifier.VerifyNs(fakeTxID, tx1, 0))
+
+	tx1.Endorsements = []*applicationpb.Endorsements{{
+		EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{},
+	}}
+	err = nsVerifier.VerifyNs(fakeTxID, tx1, 0)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "no endorsements provided")
+}
+
+func TestNsVerifierInvalidIdentities(t *testing.T) {
+	t.Parallel()
+	pItem, _ := policy.MakePolicyAndNsEndorser(t, "1")
+	pol := &applicationpb.NamespacePolicy{}
+	require.NoError(t, proto.Unmarshal(pItem.Policy, pol))
+	nsVerifier, err := signature.NewNsVerifier(pol, nil)
+	require.NoError(t, err)
+
+	for _, tc := range []struct {
+		name     string
+		identity *msppb.Identity
+	}{
+		{name: "nil identity", identity: nil},
+		{name: "empty identity struct", identity: &msppb.Identity{}},
+		{name: "identity with only MspId", identity: &msppb.Identity{MspId: "hello"}},
+		{name: "identity with unknown MspId and cert", identity: msppb.NewIdentity("hello", []byte("rand"))},
+		{name: "identity with empty MspId and cert", identity: msppb.NewIdentity("", []byte("rand"))},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			tx := &applicationpb.Tx{
+				Namespaces: []*applicationpb.TxNamespace{{
+					NsId:       "1",
+					NsVersion:  1,
+					ReadWrites: []*applicationpb.ReadWrite{{Key: []byte("k1"), Value: []byte("v1")}},
+				}},
+				Endorsements: []*applicationpb.Endorsements{{
+					EndorsementsWithIdentity: []*applicationpb.EndorsementWithIdentity{{
+						Endorsement: []byte("sign"),
+						Identity:    tc.identity,
+					}},
+				}},
+			}
+			err := nsVerifier.VerifyNs(fakeTxID, tx, 0)
+			require.Error(t, err)
+			require.Contains(t, err.Error(), "signature mismatch")
+		})
+	}
 }
 
 func TestNsVerifierPolicyBased(t *testing.T) {

